### DISCUSSION TITLE
How to share scripts between projects in a workspace?

### DISCUSSION THREAD

**Discussion by zkochan** [pnpm/pnpm#3166] [Ideas]
In the pnpm monorepo, we copy/paste the same scripts to almost every project. There should be a better way to do this.

Yarn solved this with constraints. For instance, this is how they generate the prepack scripts:

https://github.com/yarnpkg/berry/blob/20a3fdfe6ef88bf611a5b3af21abe24c86add204/constraints.pro#L62-L69

We don't have constraints support though. And seems like not everyone is currently happy with Yarn's constraints, as they are written in Prolog.

Maybe we could come up with something like a "baseScripts", which are inherited by each workspace project.

related: #1813

**Comment by muuvmuuv** [comment]
Oh this is a great idea, just came here to open a similar case. We are switching form monolith to monorepo currently because we will build multiple ionic-capacitor apps in one project (custom branded for our customers). At the moment I am using npm-run-all in the root with `"build": "run-p build:*"` and have 10 build scripts `"build:<project-name>": "pnpm run -C projects/<project-name> build"`. Just doing something like `pnpm run -w build` would be awesome so it will search in the workspace packages for a matching script and run it.

**Reply by muuvmuuv** [reply]
Very old issue but we moved all scripts to Moon and as tasks. Way faster and easy caching. https://moonrepo.dev/

**Comment by turnerguo** [comment]
I think this feature is very useful, imaging that you have 20 package in monorepo, each package have 5 commands in script such as "dev, build, lint, clean, watch", and each of them are same command just in different package, so can we just support "extends" method just like tsconfig.json did? that gives more flexibility to developer for big monorepo.

**Reply by gluxon** [reply]
We had a solid attempt at a `package.json` field sharing mechanism recently in https://github.com/pnpm/rfcs/pull/3. A few pnpm contributors thought about how this would work in a secure way for a few months.

We ultimately paused the approach after multiple reviewers were concerned about `package.json` files consuming or inheriting other `package.json` files. One example review: https://github.com/pnpm/rfcs/pull/3#issuecomment-1623770330

Linking the above in hopes it's useful to others. It's possible others can go propose something different that works better in the future.

**Comment by WandererXII** [comment]
What about adding `scripts` section to `pnpm-workspaces.yaml`, so we get something like this:

```YAML
packages:
  - 'ui/**/*'

catalog:
  snabbdom: ^3.5.1
  '@badrap/result': ^0.2

scripts:
  type-check: 'tsc --noEmit'
  lint: 'eslint . --ext .ts'
  ```
  
  Each workspace `package.json` could then have scripts listed like this:
  
  ```JSON
  "scripts": {
    "lint": "scripts:",
    "type-check": "scripts:"
  },
  ```
  Scripts would still need to be listed in each workspace, but it has similar advantages as catalogs and the same syntax.

**Comment by ClementValot** [comment]
Yarn has a not-so-documented feature where scripts with a `:` are available globally in the whole monorepo. This allows some neat possibilities like sharing a build script, e.g. all my ts packages build the same ways so I have: 

root package.json
```json
"scripts": {
    "g:build-ts": "cd $INIT_CWD && rimraf lib types && tsc"
}
```

packageA/package.json
```json
"scripts": {
    "build": "yarn g:build-ts"
}
```

Other examples include chucking all the test dependencies in a dedicated helper package which also contains a `g:test-mocha` script which is a lifesaver in terms of reducing the number of packages declarations and copy-pasted scripts. 

While this can be somewhat reproduced with `pnpm -w run g:build-ts`, I run into https://github.com/pnpm/pnpm/issues/7042
